lib/delayed_job/threaded_worker.rb

```diff
@@ -0,0 +1,104 @@
+require 'socket'
+
+class ThreadedWorker < Delayed::Worker
+  def initialize(thread_count, options={})
+    super(options)
+    @thread_count = thread_count
+    @threads = []
+    @stop_signal = false
+    @name_prefix = options[:worker_name] || 'worker'
+  end
+
+  def start
+    trap_signals
+
+    say "Starting multi-threaded job worker with #{@thread_count} threads"
+
+    @thread_count.times do |i|
+      thread_name = generate_thread_name(i + 1)
+      @threads << Thread.new do
+        thread_name(thread_name)
+        threaded_work_off
+      end
+    end
+
+    @threads.each(&:join) # Wait for threads to finish
+  end
+
+  def name
+    Thread.current[:name] || @name
+  end
+
+  def name=(name)
+    # Set the instance variable for compatibility && ensure thread-local storage is also updated
+    @name = name
+    Thread.current[:name] = name
+  end
+
+  private
+
+  def generate_thread_name(thread_index)
+    if @name
+      "#{@name}-thread:#{thread_index}"
+    else
+      "#{@name_prefix}-host:#{Socket.gethostname} pid:#{Process.pid} thread:#{thread_index}"
+    end
+  end
+
+  def thread_name(name)
+    # Store the name in thread-local storage && Set the name in the instance variable for compatibility
+    Thread.current[:name] = name
+    self.name = name
+  end
+
+  def stop_signal?
+    @stop_signal
+  end
+
+  def threaded_work_off
+    retry_attempts = 0
+
+    until stop_signal?
+      begin
+        runtime = Benchmark.realtime do
+          @result = work_off(100) # Attempt to process one job
+        end
+
+        count = @result[0] + @result[1]
+
+        if count.zero?
+          sleep(self.class.sleep_delay) unless stop_signal?
+        else
+          say sprintf("#{count} jobs processed at %.4f j/s, %d failed", count / runtime, @result.last)
+        end
+      rescue StandardError => e
+        say "Thread #{name} encountered an error: #{e.message}. Restarting thread..."
```

Comment:
**Suggestion** This isn't actually restarting the thread right? Perhaps change the phrasing here to `Thread #{name} encountered an error: #{e.message}. Retrying...`